packages/eslint-plugin/src/rules/no-floating-promises.ts

```diff
@@ -171,13 +197,17 @@ export default util.createRule<Options, MessageId>({
     function isUnhandledPromise(
       checker: ts.TypeChecker,
       node: TSESTree.Node,
-    ): boolean {
+    ): { isUnhandled: boolean; specialReason?: SpecialReasons } {
```

Comment:
[Nitpick]

> `specialReason?: SpecialReasons`

Nit: I think I see why you made the `SpecialReasons` type (leaving the code open to more reasons in the future), but for now it's not really necessary to. We can always expand to more cases later. But in the meantime it's extra cognitive complexity for a feature we aren't 100% likely to need. I'd prefer having a simpler type if possible.

```suggestion
    ): { isUnhandled: boolean; nonFunctionHandler?: boolean } {
```

Reply:
Makes sense. Changed